tianshou/data/collector.py

```diff
@@ -220,198 +298,299 @@ def collect(
+    ) -> np.ndarray:
+        """Remove surplus env id from ready_env_ids to record at least one episode per worker."""
+        surplus_env_num = len(ready_env_ids) - (n_episode - episode_count)
+        if surplus_env_num > 0:
+            mask = np.ones_like(ready_env_ids, dtype=bool)
+            mask[env_ind_local[:surplus_env_num]] = False
+            ready_env_ids = ready_env_ids[mask]
+            self.data = self.data[mask]
+        return ready_env_ids
+
+    def sample_equal_episodes_per_worker_postprocessing_on_done_env(
+        self,
+        gym_reset_kwargs: dict[str, Any] | None,
+        ready_env_ids: np.ndarray,
+        unfinished_ind_local: np.ndarray,
+    ) -> np.ndarray:
+        """Sample the same number of episodes from each worker."""
+        ready_env_ids = ready_env_ids[unfinished_ind_local]
+        self.data = self.data[unfinished_ind_local]
+        if len(unfinished_ind_local) == 0:
+            ready_env_ids = np.arange(self.env_num)
+            self._reset_env_with_obs_next_to_obs(gym_reset_kwargs)
+            for i in ready_env_ids:
+                self._reset_state(i)
+        return ready_env_ids
+
+    def step_env_and_update_data(
+        self,
+        action_remap: np.ndarray,
+        ready_env_ids: np.ndarray,
+    ) -> np.ndarray:
+        """Step the environment one action and update the data accordingly. Return the array of workerwise done signal."""
+        obs_next, rew, terminated, truncated, info = self.env.step(
+            action_remap,
+            ready_env_ids,
         )
+        done = np.logical_or(terminated, truncated)
+        self.data.update(
+            obs_next=obs_next,
+            rew=rew,
+            terminated=terminated,
+            truncated=truncated,
+            done=done,
+            info=info,
+        )
+        if self.preprocess_fn:
+            self.data.update(
+                self.preprocess_fn(
+                    obs_next=self.data.obs_next,
+                    rew=self.data.rew,
+                    done=self.data.done,
+                    info=self.data.info,
+                    policy=self.data.policy,
+                    env_id=ready_env_ids,
+                    act=self.data.act,
+                ),
+            )
+        return done
+
+    def compute_next_action_in_active_workers(
+        self,
+        last_state: Any,
+        no_grad: bool,
+        random: bool,
+        ready_env_ids: np.ndarray,
+    ) -> None:
+        """Compute the next action in the active workers and update the data accordingly."""
+        if random:
+            try:
+                act_sample = [self._action_space[i].sample() for i in ready_env_ids]
+            except TypeError:  # envpool's action space is not for per-env
+                act_sample = [self._action_space.sample() for _ in ready_env_ids]
+            act_sample = self.policy.map_action_inverse(act_sample)  # type: ignore
+            self.data.update(act=act_sample)
```

Comment:
From the method name I wouldn't expect it to update data. Generally, data is modified inplace all over the class, we should avoid it where possible